draft-ietf-tls-esni.md

```diff
@@ -341,27 +341,39 @@ ignore the `ECHConfig`.
 
 # The "encrypted_client_hello" Extension {#encrypted-client-hello}
 
-The encrypted ClientHelloInner is carried in an "encrypted_client_hello"
-extension, defined as follows:
+To offer ECH, the client sends an "encrypted_client_hello" extension in the
+ClientHelloOuter. When it does, it MUST also send the extension in
+ClientHelloInner.
 
 ~~~
     enum {
        encrypted_client_hello(0xfe0a), (65535)
     } ExtensionType;
 ~~~
 
-When offered by the client, the extension appears only in the ClientHelloOuter.
-The payload MUST have the following structure:
+The payload of the extension has the following structure:
 
 ~~~~
+    enum { outer(0), inner(1) } ClientECHType;
+
     struct {
-       HpkeSymmetricCipherSuite cipher_suite;
-       uint8 config_id;
-       opaque enc<1..2^16-1>;
-       opaque payload<1..2^16-1>;
+       ClientECHType type;
+       select (ClientECH.type) {
+           case outer:
+               HpkeSymmetricCipherSuite cipher_suite;
+               uint8 config_id;
+               opaque enc<1..2^16-1>;
+               opaque payload<1..2^16-1>;
+           case inner:
+               Empty;
+       };
     } ClientECH;
 ~~~~
 
+The outer extension uses the `outer` variant and the inner extension uses the
+`inner` variant. The inner extension has an empty payload. The outer
+extension has the following fields:
+
 config_id
 : The ECHConfigContents.key_config.config_id for the chosen ECHConfig.
 
```

Comment:
(This comment should be a few stanzas down, but GitHub is silly.)

Something I realized working through an implementation. I think we also need to adjust this line below:

>  When the client offers the "encrypted_client_hello" extension, the server MAY include an "encrypted_client_hello" extension in its EncryptedExtensions message with the following payload:

If we handshake a ClientHelloInner, the server is not supposed to send retry configs. The encrypted_client_hello / ech_is_inner version captured this implicitly with the unsolicited extensions rule: the inner ClientHello doesn't have encrypted_client_hello, only ech_is_inner.

In this form, we instead make HRR happier, but I think we need to here, or elsewhere, say, the server can only send this when the client offers encrypted_client_hello with type outer. Sending this with type inner should be an unsupported_extension alert.

Reply:
Done.